### DISCUSSION TITLE
Windows Explorer Details (Meta Data, Exif, Attributes) -

### DISCUSSION THREAD

**Discussion by page2pagepro** [audiamus/BookLibConnect#96] [Ideas]
Below is a VBScript file that will dump relevant AudioBook details from the files:
```VBScript
Const ShowSelectRelevant = False ' True is relevant, False is all data 0 thru iDetails
Const iDetails = 321 ' Nothing found after this index
Dim aHeaders()
ReDim aHeaders(iDetails)

Set oSHL = CreateObject("Shell.Application")
Set oFLD = oSHL.Namespace("N:\BookLib")

For i = 0 to iDetails
	aHeaders(i) = oFLD.GetDetailsOf(oFLD.Items, i)
Next

If ShowSelectRelevant Then
	Dim sDetails, aDetails
	sDetails = "0,1,2,3,4,13,14,15,20,21,24,27,28,164,165,237"
	aDetails = Split(sDetails, ",")
	For Each strFileName in oFLD.Items
		For i = LBound(aDetails) To UBound(aDetails)
			Wscript.Echo aDetails(i) & vbtab & aHeaders(aDetails(i)) & ": " & oFLD.GetDetailsOf(strFileName, aDetails(i))
		Next
		WScript.Echo Replace(Space(80), " ", "-")
	Next
Else
	For Each strFileName in oFLD.Items
		For i = 0 to iDetails
			Wscript.Echo i & vbtab & aHeaders(i) & ": " & oFLD.GetDetailsOf(strFileName, i)
		Next
		WScript.Echo Replace(Space(80), " ", "-")
	Next
End If

Set oFLD = Nothing
Set oSHL = Nothing
```

I notice a duplication of efforts:

- (014) Album = (021) Title
- (013) Contributing artists = (020) Authors

Fields that exist in SQL "%LOCALAPPDATA%\audiamus\BookLibConnect\data\audiobooks.db" I'd hope to convey to the flat-file level include:

- Books.Subtitle -> Perhaps to (014) Album
- Books. -> Perhaps to (243) Composers

Using a quick SQLite statement below we can gather in a single view the series, title, author(s), and narrator(s):
```SQL
SELECT
	Books.Subtitle As 'Series', Books.Title, Authors.Name As 'Authors(s)', Narrators.Name As 'Narrator(s)'
FROM
	Books, Authors, AuthorBook, BookNarrator, Narrators
WHERE
	Books.Id = AuthorBook.BooksId AND 
	AuthorBook.AuthorsId = Authors.Id AND
	BookNarrator.NarratorsId = Narrators.Id AND
	BookNarrator.BooksId = Books.Id
```

Using EXIFTOOL, the Narrator does exist properly, meaning the data is embedded, just not in a fashion Microsoft Explorer can understand.

As I make progress, I may build a small middleware to add/modify meta data to suite my needs.

Please feel free to add to the discussion any suggestions or directives you'd like realized.

Thank you kindly!

**Comment by audiamus** [comment]
Let me explain some context:

Book Lib Connect does not touch meta data at all. This is by intention. Its sole purpose is to download the Mpeg4 container - which is quite a task in itself - and decrypt the AAC audio within, using excellent [AAXClean](https://github.com/Mbucari/AAXClean), which is based on relevant components of FFmpeg. The result will be saved with the .m4b extension, but other than being plain audio, it's exactly the same as Audible has uploaded it. The only modification that Book Lib Connect makes is to flatten the separate json meta for chapters which it also downloads from the Audible servers. 

The reason for not doing anything else is the existence of [AAX Audio Converter](https://github.com/audiamus/AaxAudioConverter). Book Lib Connect is a rather new companion to AAX Audio Converter. It was needed when Audible retired their Win10 app which previously served as the downloader for AAX Audio Converter.

AAX Audio Converter does all the real conversion work, and also processes the meta data. AAX Audio Converter first extracts all meta data from the Mpeg4 container, then completely removes it, including the Audible proprietary tags. It then rebuilds meta data from scratch, delegating this to marvellous [ATL .Net](https://github.com/Zeugma440/atldotnet), only using "standard" tags. Tags applied for persons, i.e author and narrator, can be customized in the AAX Audio Converter settings.
